### PR TITLE
Authn: Handle logout logic in auth broker

### PR DESCRIPTION
**Pull request by kalleep**
**What is this feature?**
The logout handler previously had a bunch of provider specific logic when performing logout. This part was a bit messy and hard to follow. 

I instead added logout functionality to `authn.Service`. The service it self perform all shared logic, revoking session token and generate default redirect url.

Some clients (oauth and saml) can be configured with additional logic that should happen. To support this I added a extensions interface that clients can implement `authn.LogoutClient`. If a client implements this interface it will be called for users that has authenticated with it.

For now it is only implemented for oauth clients. For these we remove access, refresh and id tokens and if configured create a new redirect url with support for [oidc rp initiated logout](https://openid.net/specs/openid-connect-rpinitiated-1_0.html).

Plan is to also fix enterprise saml client to also implement this interface and remove the TODO added here. 

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/identity-access-team/issues/496

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/authn/authntest/fake.go
Comment (gamab) [inline pkg/services/authn/authntest/fake.go:72]: TODO I guess?

File: pkg/services/authn/authn.go
Comment (gamab) [inline pkg/services/authn/authn.go:78]:
```suggestion
	// Logout revokes session token and does additional clean up if client used to authenticate supports it
```

File: pkg/services/authn/authn.go
Comment (gamab) [inline pkg/services/authn/authn.go:121]:
```suggestion
// LogoutCLient is an optional interface that auth client can implement.
```

File: pkg/services/authn/clients/oauth.go
Comment (gamab) [inline pkg/services/authn/clients/oauth.go:229]: Having done exactly that a few days ago, I think you sadly need to do it in two steps if `token.Extra("id_token")` is `nil` then that panics. Then IMO, it's safe to assume that if it's set then it's a string.
```suggestion
		if idToken := token.Extra("id_token"); idToken != nil {
			redirctURL = withIDTokenHint(redirctURL, idToken.(string))
```

File: pkg/services/authn/clients/oauth_test.go
Comment (gamab) [inline pkg/services/authn/clients/oauth_test.go:363]: Duplicated test case
```suggestion
```

File: pkg/services/authn/authnimpl/service.go
Comment (gamab) [inline pkg/services/authn/authnimpl/service.go:347]: Seems like we added traces in other functions, do you want to add a trace?

**Review by gamab — COMMENTED** [review]
This is truly amazing! Very nice and clean refactor.
I just want to give it a quick test before a final approval.

File: pkg/services/authn/authntest/fake.go
Comment (kalleep) [reply pkg/services/authn/authntest/fake.go:72]: I think leaving it like this in a fake is fine

File: pkg/services/authn/clients/oauth.go
Comment (gamab) [reply pkg/services/authn/clients/oauth.go:229]: I was wrong: 
```go
func TestService_Token(t *testing.T) {
	token := &oauth2.Token{
		AccessToken:  "aa.bb.cc",
		Expiry:       time.Now(),
		RefreshToken: "bb.cc.dd",
		TokenType:    "blabla",
	}

	_, ok := token.Extra("id_token").(string)

	assert.False(t, ok)
}
```

**Review by gamab — APPROVED** [review]
Tested with AzureAD OIDC Logout, SAML logout, Generic OAuthLogout (with no logout url). Everything worked fine :)

**Pull request merged**
